### ISSUE TITLE
AWS Sync fails with TypeError when being called from library context.

### ISSUE THREAD

**Issue by trega123** [aws/aws-cli#3358]
Hello AWS,

We are using the awscli from a python import and have found that the AWS sync functionality erroneously makes the expectation that the LastModified value is always a string and attempts to cast to a datetime. This functionality works when running the code from the context of a command line because datetimes are cast to a string; however, when running this code with the default boto3 behavior, LastModified is a datetime and it will fail.

See the code snippet below for a way to reproduce this problem. This code executes as an AWS lambda with the entry point of lambda_handler.:

```
from awscli.clidriver import CLIDriver
from awscli.customizations.s3.s3 import S3

class NewS3CLIDriver(CLIDriver):
    """Returns Amazon's CLI Driver with the new S3 plugin
    registered as s3 and the old plugin registered as s3api."""

    def _get_command_table(self):
        """Overrides the default get command table behavior to register the s3 plugin."""
        if self._command_table is None:
            super(NewS3CLIDriver, self)._get_command_table()
            self._command_table["s3api"] = self._command_table["s3"]
            self._command_table["s3"] = S3(self.session)

        return self._command_table

def lambda_handler(event, context):
    """Executes the AWS s3 sync command"""
    driver = NewS3CLIDriver()

    driver.main(["s3", "sync", "s3://b1/", "s3://b2/"])
```

Here is the traceback for the error raised:

```
Traceback (most recent call last):
  File "/var/task/awscli/customizations/s3/s3handler.py", line 173, in call
    for fileinfo in fileinfos:
  File "/var/task/awscli/customizations/s3/fileinfobuilder.py", line 31, in call
    for file_base in files:
  File "/var/task/awscli/customizations/s3/comparator.py", line 79, in call
    src_file = advance_iterator(src_files)
  File "/var/task/awscli/customizations/s3/filegenerator.py", line 142, in call
    for src_path, extra_information in file_iterator:
  File "/var/task/awscli/customizations/s3/filegenerator.py", line 324, in list_objects
    extra_args=extra_args):
  File "/var/task/awscli/customizations/s3/utils.py", line 375, in list_objects
    content['LastModified'])
  File "/var/task/awscli/customizations/s3/utils.py", line 351, in _date_parser
    return parse(date_string).astimezone(tzlocal())
  File "/var/runtime/dateutil/parser/_parser.py", line 1312, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/var/runtime/dateutil/parser/_parser.py", line 604, in parse
    res, skipped_tokens = self._parse(timestr, **kwargs)
  File "/var/runtime/dateutil/parser/_parser.py", line 680, in _parse
    l = _timelex.split(timestr)         # Splits the timestr into tokens
  File "/var/runtime/dateutil/parser/_parser.py", line 207, in split
    return list(cls(s))
  File "/var/runtime/dateutil/parser/_parser.py", line 76, in __init__
    '{itype}'.format(itype=instream.__class__.__name__))
TypeError: Parser must be a string or character stream, not datetimefatal error: Parser must be a string or character stream, not datetime
```

I have opened a pull request (#3357) which resolves this issue and includes a unit test.

Thanks, and let me know if you have any questions.
Thomas Rega
Cox Automotive Inc.

**Comment by kyleknap** [comment]
Thanks for the PR. One thing that I am concerned about is that the internals of the CLI is not meant to be used publicly as we routinely break its internal interface. Is there a specific reason, you are not just installing the CLI with `pip` and calling the executable from a subprocess? You should be able to do that and that interface is guaranteed not to break.

**Comment by trega123** [comment]
Hello Kyle,

Our use case is to run an S3 sync operation from an AWS lambda. Boto3 does not include the S3 sync functionality as it is a plugin for the aws-cli, otherwise we would have used boto3. The only resource we found to run the S3 sync operation from inside of a lambda is here:

[https://alestic.com/2016/11/aws-lambda-awscli/](url)

This would fit our needs; however, I would have preferred the approach of not shelling out using subprocess, especially from a lambda function on a machine we do not control. It is also a pain to setup the environment as aws-cli is not available by default in lambda and neither is pip. This is why I decided to instead import aws-cli from inside python and call the main method on the CLIDriver class directly.

I understand your concern with the internals of aws-cli changing without notice, but I guess that is a risk I am willing to take! Thanks for your feedback, and let me know if you have any suggestions for how to better improve this process.
